From 80e134d155d81cda91e1d2d471a35bd069583a84 Mon Sep 17 00:00:00 2001
From: fwaters <fwaters@anaconda.com>
Date: Mon, 23 Sep 2019 18:34:40 +0000
Subject: [PATCH 5/7] intel_setup_add_whl_deps

---
 setup.py | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/setup.py b/setup.py
index e6e696c11..e48c7373d 100755
--- a/setup.py
+++ b/setup.py
@@ -62,6 +62,9 @@ MICRO               = 1
 ISRELEASED          = True
 VERSION             = '%d.%d.%d' % (MAJOR, MINOR, MICRO)
 
+LOCAL               = 'intel.%s' % os.environ.get('PKG_BUILDNUM','0')
+ISLOCAL             = 'bdist_wheel' in sys.argv[1:]
+
 
 # Return the git revision as a string
 def git_version():
@@ -123,6 +126,8 @@ def get_version_info():
 
     if not ISRELEASED:
         FULLVERSION += '.dev0+' + GIT_REVISION[:7]
+    else:
+        FULLVERSION += '+%s' % LOCAL
 
     return FULLVERSION, GIT_REVISION
 
@@ -145,7 +150,7 @@ if not release:
 
     a = open(filename, 'w')
     try:
-        a.write(cnt % {'version': VERSION,
+        a.write(cnt % {'version': FULLVERSION if ISLOCAL else VERSION,
                        'full_version': FULLVERSION,
                        'git_revision': GIT_REVISION,
                        'isrelease': str(ISRELEASED)})
@@ -452,6 +457,8 @@ def setup_package():
             'console_scripts': f2py_cmds
         },
     )
+    if ISLOCAL:
+        metadata["install_requires"] = [ 'icc_rt','mkl','mkl_fft','mkl_random','tbb4py' ]
 
     if "--force" in sys.argv:
         run_build = True
-- 
2.20.1

